### PR TITLE
gh-108303: Move `Lib/test/sortperf.py` to `Tools/scripts`

### PR DESCRIPTION
**Pull request by sobolevn**
@corona10 what do you think about moving these case to `pyperformance`? Would it be benefitial?

* Issue: gh-108303

### CODE REVIEW

**Comment by sobolevn** [comment]
It was Guido 27 years ago: https://github.com/python/cpython/commits/main/Lib/test/sortperf.py

I think he agreed on moving it away in Discord's #python-dev channel.

**Review by corona10 — CHANGES_REQUESTED** [review]
IMO, the First thing we have to do is add the benchmark script, not remove the file.

**Comment by sobolevn** [comment]
@corona10 ok, I will send a PR to pyperformance soon! 👍

**Review by corona10 — COMMENTED** [review]
Even if we just move the script into Tools/script 
IMO pyperf based script is more modernized way and easy to compare with improvement patch. WDYT?

**Comment by sobolevn** [comment]
I agree with Guido: it is not really that useful as a benchmark (since it is a micro-benchmark). It is only interesting when working on `list.sort` directly. 

So, I think that we can move this file for now. If someone has a real use-case for `list.sort` benchmark in `pyperf`, they can recreate my existing closed PR.

**Comment by corona10** [comment]
> So, I think that we can move this file for now. If someone has a real use-case for list.sort benchmark in pyperf, they can recreate my existing closed PR.

Ah, I meant rewriting the benchmark script based on pyperf(you already did) and storing it Tools/script/
I also agree with Guido.

**Comment by sobolevn** [comment]
@corona10 sure!

**Comment by corona10** [comment]
I will take a look at the PR by tomorrow :)

File: Tools/scripts/sortperf.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    def _prepare_data(self, loops: int) -> list[float]:
+        bench = BENCHMARKS[self._name]
+        return [
+            bench(self._size, self._random)
+            for _ in range(loops)
+        ]
+
+
+def add_cmdline_args(cmd: list[str], args) -> None:
+    cmd.append(args.benchmark)
```
Comment (corona10) [inline Tools/scripts/sortperf.py]:
```suggestion
def add_cmdline_args(cmd: list[str], args) -> None:
    if args.benchmark:
        cmd.append(args.benchmark)
```

File: Tools/scripts/sortperf.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    cmd.append(f"--size={args.size}")
+    cmd.append(f"--rng-seed={args.rng_seed}")
+
+
+def add_parser_args(parser: argparse.ArgumentParser) -> None:
+    parser.add_argument(
+        "benchmark",
+        choices=BENCHMARKS,
+        nargs="?",
+        default="list_sort",
```
Comment (corona10) [inline Tools/scripts/sortperf.py]:
```suggestion
```

File: Tools/scripts/sortperf.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    runner = pyperf.Runner(add_cmdline_args=add_cmdline_args)
+    add_parser_args(runner.argparser)
+    args = runner.parse_args()
+
+    runner.metadata["description"] = "Test `list.sort()` with different data"
+    runner.metadata["list_sort_size"] = args.size
+    runner.metadata["list_sort_random_seed"] = args.rng_seed
+
+    benchmark = Benchmark(args.benchmark, args.size, args.rng_seed)
+    runner.bench_time_func(args.benchmark, benchmark.run)
```
Comment (corona10) [inline Tools/scripts/sortperf.py]:
```suggestion
    if args.benchmark:
        benchmarks = (args.benchmark,)
    else:
        benchmarks = sorted(BENCHMARKS)
    for bench in benchmarks:
        benchmark = Benchmark(bench, args.size, args.rng_seed)
        runner.bench_time_func(bench, benchmark.run)
```

**Review by corona10 — CHANGES_REQUESTED** [review]
1. By default, all benchmark should be executed.
2. Benchmarks needs some tune up
```
.....................
list_sort: Mean +- std dev: 30.8 ms +- 0.4 ms
.....................
list_sort_ascending: Mean +- std dev: 1.76 ms +- 0.10 ms
```

Execution time is too long :(
(Maybe we need to control inner_loops value)

**Comment by sobolevn** [comment]
@corona10 I've merged your suggestions. Thank you!

I've also optimized the benchmark:

```
» time python Tools/scripts/sortperf.py  
.....................
list_sort: Mean +- std dev: 29.0 us +- 0.2 us
.....................
list_sort_ascending: Mean +- std dev: 28.9 us +- 0.4 us
.....................
list_sort_ascending_exchanged: Mean +- std dev: 29.2 us +- 0.4 us
.....................
list_sort_ascending_one_percent: Mean +- std dev: 29.5 us +- 0.2 us
.....................
list_sort_ascending_random: Mean +- std dev: 29.6 us +- 0.3 us
.....................
list_sort_descending: Mean +- std dev: 29.5 us +- 0.2 us
.....................
list_sort_duplicates: Mean +- std dev: 30.4 us +- 1.1 us
.....................
list_sort_equal: Mean +- std dev: 29.2 us +- 0.2 us
.....................
list_sort_worst_case: Mean +- std dev: 29.2 us +- 0.2 us
python Tools/scripts/sortperf.py  98.38s user 3.21s system 99% cpu 1:41.75 total
```

Is it good enough?

**Review by corona10 — APPROVED** [review]
LGTM

@sobolevn Feel free to merge the PR anytime!

**Comment by sobolevn** [comment]
@corona10 Thank you! 🎉 


I think that we also need to backport this to 3.11 and 3.12
Do you agree?

**Comment by corona10** [comment]
Yeah go a head

**Pull request merged**